capirca/lib/fortigate.py

```diff
@@ -0,0 +1,383 @@
+      except KeyError:
+        raise FortiGatePortDoesNotExist
+    else:
+      raise FortiGateFindServiceError(
+        'failed to get service from %r protocol and %r port' % (protocol, port)
+      )
+
+
+class ObjectsContainer:
+  """a Container that holds service and network objects"""
+
+  def __init__(self):
+    self._FW_ADDRESSES = []
+    self._FW_SERVICES = []
+
+    self._FW_DUP_CHECK = set()
+
+  def get_fw_addresses(self):
+    """return the collected addresses"""
+    self._FW_ADDRESSES.extend([' ', 'end', ' '])
+    return self._FW_ADDRESSES
+
+  def get_fw_services(self):
+    """return the collected services"""
+    self._FW_SERVICES.extend([' ', 'end', ' '])
+    return self._FW_SERVICES
+
+  def _add_address_to_fw_addresses(self, addr):
+    """add address to address store"""
+    if addr in self._FW_DUP_CHECK:
+      return
+    self._FW_ADDRESSES.extend(['\tedit %s' % addr,
+                               '\t\tset subnet %s' % addr,
+                               '\tnext'])
+    self._FW_DUP_CHECK.add(addr)
+
+  def _add_service_to_fw_services(self, protocol, service):
+    """add service to services store"""
+    if service in self._FW_DUP_CHECK:
+      return
+
+    self._FW_SERVICES.extend(
+      ['\tedit %s' % service,
+       '\t\tset protocol TCP/UDP',
+       '\t\tset %s-portrange %s' % (protocol.lower(), service),
+       '\tnext']
+    )
+
+    self._FW_DUP_CHECK.add(service)
+
+
+class Term(aclgenerator.Term):
+  """Single Firewall Policy"""
+  ALLOWED_PROTO_STRINGS = ['gre', 'icmp', 'ip', 'tcp', 'udp']
+  COMMENT_MAX_WIDTH = 70
+
+  CURRENT_ID = 0
+
+  def __init__(self, term, object_container):
+    super(Term, self).__init__(term)
+    self._term = term
+    self._obj_container = object_container
+
+    self.id_ = type(self).CURRENT_ID
+    type(self).CURRENT_ID += 1
+
+  @staticmethod
+  def _get_addresses_name(addresses):
+    """return the addresses or 'all' if no addresses specified"""
+    v4_addresses = [x.with_prefixlen for x in addresses if
+                    not isinstance(x, nacaddr.IPv6)]
+    addresses = ' '.join(v4_addresses)
+    return addresses or 'all'
+
+  @staticmethod
+  def clean_ports(src_ports, dest_ports):
+    """return a set() of src and dest ports"""
+    all_ports = []
+    if src_ports:
+      all_ports += src_ports
+    if dest_ports:
+      all_ports += dest_ports
+    return set(all_ports)
+
+  def _get_services_string(self, protocol, ports):
+    """
+    get the service name if exist, if not create a service object and return the name
+    :param protocol: list of protocols
+    :param ports: list of ports
+    :return:
+    """
+
+    services = []
+    if protocol and not ports:
+      services.append(FortigatePortMap.GetProtocol(protocol[0]))
+    for port in ports:
+      try:
+        service = FortigatePortMap.GetProtocol(protocol[0], port[0])
+      except FortiGatePortDoesNotExist:
+        self._obj_container._add_service_to_fw_services(protocol[0], port[0])
```

Comment:
you are accessing a protected method of another object. This usually means that this method shouldn't be protected or that you shouldn't be doing that.